datalad/support/operations/operations_abstract.py

```diff
@@ -0,0 +1,192 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+"""
+from datalad.utils import Path
+
+
+class OperationsBase(object):
+    """Abstract class with the desired API for operations.
+
+    Any path-like argument must be a pathlib object.
+
+    Method parameters must be verbose or generic enough to be translatable
+    to non-POSIX systems, such as Windows.
+    """
+    def __init__(self, cwd=None):
+        """
+        Parameters
+        ----------
+        cwd : Path or None
+          Current working directory to resolve any relative paths against.
+          If None, the process working directory (PWD) is used.
+        """
+        self._cwd = Path.cwd() if cwd is None else cwd
+
+    def _ensure_absolute(self, path):
+        """Internal helper to make a path absolute
+
+        If relative path is given, interpret it as relative to initialized CWD
+        if any, to PWD otherwise.
+
+        Parameters
+        ----------
+        path : Path
+
+        Returns
+        -------
+        Path
+        """
+
+        # Note, that this doesn't actually resolve the path. For now, this is
+        # intentional to not get completely confusing return values. This is
+        # only about interpreting as relative to the correct base.
+        # TODO: possibly turn this into a decorator, that allows to assign
+        #       arguments of the actual methods, that need to go through this.
+        if not path.is_absolute():
+            if self._cwd:
+                path = self._cwd / path
+            else:
+                path = Path.cwd() / path
+        return path
+
+    def make_directory(self, path, force=False):
+        """Create a new directory
+
+        Parameters
+        ----------
+        path : Path
+          Location at which to create a directory.
+        force : bool
+          Enforce creation of directory, even when it already exists,
+          or if parent directories are missing.
+        """
+        raise NotImplementedError
+
+    def exists(self, path):
+        """Test if something exists at the given path
+
+        Parameters
+        ----------
+        path : Path
+          Path to test.
+
+        Returns
+        -------
+        bool
+          True, if the path exists (even in case of a broken symlink), False
+          otherwise.
+        """
+        raise NotImplementedError
```

Comment:
`listdir` or alike? then could be used in `create_sibling` safeguard: https://github.com/datalad/datalad/pull/4147/files#diff-865501de3883a6b49b32476bdc7b3380R305

Reply:
Yes, agree.